File: app/src/main/java/com/jx/androiddemo/testactivity/function/f31to40/f32/F32Activity.java
package com.jx.androiddemo.testactivity.function.f31to40.f32;

import android.annotation.SuppressLint;
import android.os.Build;
import android.view.View;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;

import com.jx.androiddemo.BaseMvpActivity;
import com.jx.androiddemo.R;
import com.jx.androiddemo.testactivity.empty.EmptyContract;
import com.jx.androiddemo.testactivity.empty.EmptyPresenter;
import com.jx.androiddemo.tool.ClickListenerUtil;

import java.io.File;
import java.util.concurrent.TimeUnit;

import butterknife.BindView;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;

public class F32Activity extends BaseMvpActivity<EmptyPresenter> implements EmptyContract.View {

    @BindView(R.id.seekbar)
    SeekBar mSeekBar;
    @BindView(R.id.tv_current_time)
    TextView mCurrentTimeTv;
    @BindView(R.id.tv_total_time)
    TextView mTotalTimeTv;
    @BindView(R.id.img_play_pause)
    ImageView mPlayPauseImg;

    //private String mPath1 = "https://guiyu-tici.oss-cn-shanghai.aliyuncs.com/tici/149-2-20220411173608441.wav";
    private String mPath2 = "https://guiyu-tici.oss-cn-shanghai.aliyuncs.com/tici/149-2-20220331174043925.aac";
    private String mPath3 = "https://guiyu-tici.oss-cn-shanghai.aliyuncs.com/tici/149-2-20220222095449261.aac";
    private String mPath1 = "";

    @Override
    protected void initInject() {
        getActivityComponent().inject(this);
    }

    @Override
    protected int getLayout() {
        return R.layout.activity_f32;
    }

    @SuppressLint("CheckResult")
    @Override
    protected void initEventAndData() {
        initView();
        initListener();
    }

    @SuppressLint({"SetTextI18n", "CheckResult"})
    private void initView() {
        mPath1 = getExternalCacheDir() + File.separator + "f30.wav";
        AudioPlay.getInstance().setListener(new AudioPlay.Listener() {
            @SuppressLint("SetTextI18n")
            @Override
            public void onDurationChange(int duration) {
                postUi(() -> {
                    mTotalTimeTv.setText(Integer.toString(duration));
                    mSeekBar.setMax(duration);
                });
            }

            @SuppressLint("SetTextI18n")
            @Override
            public void onCurrentDurationChange(int duration) {
                postUi(() -> {
                    mCurrentTimeTv.setText(Integer.toString(duration));
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                        mSeekBar.setProgress(duration, true);
                    } else {
                        mSeekBar.setProgress(duration);
                    }
                });
            }

            @Override
            public void onPlayStatusChange(boolean isPlay) {
                postUi(() -> {
                    mPlayPauseImg.setImageResource(isPlay ? R.mipmap.img_audio_pause : R.mipmap.img_audio_play);
                });
            }
        });
        mSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {

            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
                AudioPlay.getInstance().pause();
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                AudioPlay.getInstance().playPause(mPath1, seekBar.getProgress());
            }
        });
        AudioPlay.getInstance().getAudioTotalDuration(mPath1, mSeekBar, mTotalTimeTv);
    }

    @Override
    protected void onResume() {
        super.onResume();
    }

    @Override
    protected void onPause() {
        super.onPause();
        AudioPlay.getInstance().pause();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        AudioPlay.getInstance().quit();
    }

    @SuppressLint("CheckResult")
    private void initListener() {
        //延时方法
        Observable.timer(50, TimeUnit.MILLISECONDS)
                .compose(this.bindToLifecycle())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(integer -> {

                });

        //点击
        /*RxView.clicks(null)
                .throttleFirst(Constants.CLICK_TIME, TimeUnit.MILLISECONDS)
                .compose(this.bindToLifecycle())
                .subscribe(o ->
                {
                });*/
    }

    public void onClick(View view) {
        if (!ClickListenerUtil.canClick()) return;
        switch (view.getId()) {
            case R.id.img_play_pause:
                AudioPlay.getInstance().playPause(mPath1, null);
                break;
        }
    }

    private void postUi(Runnable runnable) {
        if (null == mPlayPauseImg || null == runnable) {
            return;
        }
        mPlayPauseImg.post(runnable);
    }
}

File: app/src/main/java/com/jx/androiddemo/testactivity/function/f31to40/f32/F32Activity.kt
package com.jx.androiddemo.testactivity.function.f31to40.f32

import android.annotation.SuppressLint
import android.util.Log
import android.view.View
import android.widget.TextView
import com.jakewharton.rxbinding2.view.RxView
import com.jx.androiddemo.BaseMvpActivity
import com.jx.androiddemo.R
import com.jx.androiddemo.constant.Constants
import com.jx.androiddemo.testactivity.empty.EmptyContract
import com.jx.androiddemo.testactivity.empty.EmptyPresenter
import com.jx.androiddemo.tool.ClickListenerUtil
import kotlinx.android.synthetic.main.activity_f32.*
import java.util.concurrent.TimeUnit

class F32Activity : BaseMvpActivity<EmptyPresenter>(), EmptyContract.View {

    val mFoodProduction: Production<Food> = FoodProduction()
    val mFastFoodProduction: Production<Food> = FastFoodProduction()
    val mBurgerProduction: Production<Food> = BurgerProduction()
    val mFoodConsumer: Consumer<Burger> = FoodConsumer()
    val mFastFoodConsumer: Consumer<Burger> = FastFoodConsumer()
    val mBurgerConsumer: Consumer<Burger> = BurgerConsumer()

    companion object {
        val TAG2 = "f32"
    }

    override fun initInject() {
        activityComponent.inject(this)
    }

    override fun getLayout(): Int {
        return R.layout.activity_f32
    }

    override fun initEventAndData() {
        initView()
    }

    @SuppressLint("CheckResult")
    private fun initView() {
        val views = arrayListOf<TextView>(tv_1, tv_2, tv_3)
        var num = 0
        while (num < views.size) {
            var i = num
            views[num].setOnClickListener { v -> Log.d(TAG2, "${i}  ${num}") }
            num++
        }
        RxView.clicks(tv_4)
            .throttleFirst(Constants.CLICK_TIME.toLong(), TimeUnit.MILLISECONDS)
            .compose(this.bindToLifecycle())
            .subscribe { v ->
                mFoodProduction.produce()
                mFastFoodProduction.produce()
                mBurgerProduction.produce()
            }
        RxView.clicks(tv_5)
            .throttleFirst(Constants.CLICK_TIME.toLong(), TimeUnit.MILLISECONDS)
            .compose(this.bindToLifecycle())
            .subscribe { v ->
                val burger = Burger()
                mFoodConsumer.consume(burger)
                mFastFoodConsumer.consume(burger)
                mBurgerConsumer.consume(burger)
            }
    }

    private fun onClick(view: View) {
        if (!ClickListenerUtil.canClick()) return
        when (view.id) {
        }
    }

    class FoodProduction : Production<Food> {
        override fun produce(): Food {
            Log.d(TAG2, "produce food")
            return Food()
        }
    }

    class FastFoodProduction : Production<FastFood> {
        override fun produce(): FastFood {
            Log.d(TAG2, "produce fastfood")
            return FastFood()
        }
    }

    class BurgerProduction : Production<Burger> {
        override fun produce(): Burger {
            Log.d(TAG2, "produce burger")
            return Burger()
        }
    }

    class FoodConsumer : Consumer<Food> {
        override fun consume(item: Food) {
            Log.d(TAG2, "eat food")
        }
    }

    class FastFoodConsumer : Consumer<FastFood> {
        override fun consume(item: FastFood) {
            Log.d(TAG2, "eat fastfood")
        }
    }

    class BurgerConsumer : Consumer<Burger> {
        override fun consume(item: Burger) {
            Log.d(TAG2, "eat burger")
        }
    }
}